### DISCUSSION TITLE
PESTPP to place capture wells?

### DISCUSSION THREAD

**Discussion by RyanConway91** [usgs/pestpp#273] [Q&A]
I am looking into options to optimize capture at a well field (minimize the number of wells and total system pumping rate such that all MODPATH particles are captured). 

In addition to a couple wells that are in-place, I want to find spots and rates for new wells. This is part that I am unsure if PESTPP (OPT or MOU) can help with. I think OPT/MOU would work fine optimizing the rates of the existing wells, but I am not sure how it could be used to place some number of new wells. 

I know there are some old codes that can be used for this but I have never used them. Any ideas on tools for this problem? Is there a way to squish this into a shape that PESTPP OPT/MOU can chew on?

**Comment by jtwhite79** [comment]
this type of highly nonlinear optimization is well beyond the reach of opt...Ive done some experiments with mou and it theoretically and algorithmically has the capability seek new well locations and rates as part of an optimization.  But its computationally demanding (as are all things with mou).  

And more than that, Ive found you have to be careful and very specific about the questions you ask mou to answer.  For example I was seeking optimal injection and extraction locations and the objective was to minimize the total system rate, subject to a constraint that concentration at the down-gradient compliance point be less than X.  Well, mou figured out to put one extractor right in the center of mass at a low recovery rate and one injector immediately up-gradient of the compliance well to create a bubble of low-concentration water at the compliance point - that's evolution for you!  So if you do go down the path of optimizing locations, you'll likely need to explicitly build all those obvious things and common sense into the optimization problem (like in my example, I had to add an objective to minimize mass reaching the down-gradient boundary and also setup an exclusion zone around the compliance well). 

So its doable, but user beware!

**Reply by RyanConway91** [reply]
Thanks, good to hear mou is a potential solution. So to clarify, mou can decide how many wells to place, not just move around a specified number of wells? Whats the approach for activating/deactivating wells? Would you start too many wells and try and get mou to shut off as many as possible?

**Reply by jtwhite79** [reply]
To minimize the number of wells used in the system, you can add an objective that is the count of the number of wells with a non-zero rate (you'll have to make simple post processor to calculate this value) and ask mou to minimize that quantity. Or you can calculate the cost of drilling each well (if you know its depth and cost per linear foot and mobilization cost) and then tell mou to minimize cost (which is usually more fun and exciting for decision makers!).  

With mou, its generally good to start "feasible" (so with lots of wells and using relatively high rates to make sure all the mass/particles are being captured and/or concentrations are low at important places - assuming that output is a constraint), and then have objectives to minimize both total system rate and the number wells...or build up a single cost objective that includes installation and operation costs and tell mou to minimize (but this then can lead to a system that runs really low for a really long time, which is also not so good, so then you need an objective to minimize clean up time - see what I mean about building in common sense?!)

**Reply by RyanConway91** [reply]
Thanks, this is very helpful. I am going through the GMDSI opt/mou notebooks (thanks for those!) to get a little more familiar with optimization. Lots to figure out in those... 

One other question for getting started... Would an optimization constraint (minimization objective) based on mass (from a mf6 transport) be less non-linear and thus easier to deal with than one base on non-captured particles? Non-captured particles will be an integer whereas mass that "gets past" the well field seems more differentiable and thus better suited to PESTPP algorithms.

**Comment by jtwhite79** [comment]
I think particles would smooth enough as long as you use a lot of particles.  Mou can deal with discontinuous relations.  But you'll need to start with a population that has a fair number of members that are capturing all the particles (if that is a constraint).

**Reply by RyanConway91** [reply]
I got a MOU setup running where I am minimizing the total system pumping rate and the number of new wells, with the constraint that all particles must be captured (thanks for the input!). Would it be possible to make MOU only turn wells on and off instead of adjusting the rate? I know PEST does not do binary variables, but I could easily set the relevant decision variable parameter bounds to something like 0 to 1, and write a pre-processing script to say <0.5 = off, >=0.5 = on... Basically try and hack MOU into doing integer optimization (I think this is a version of knapsack problem?...). I am guessing this is not a good idea? The derivatives would be nonsense. I have only a few dozen potential well locations and model is pretty fast so I could do brute force without too much pain...

**Reply by jtwhite79** [reply]
That 0 to 1 trick is what Ive done also - MOU handles it without a problem because it doesn't use derivatives.  The way I set this up was similar to what you suggest with 0.5 threshold and simple model preprocessor to make the choice of on or off.  I also made "number of wells" an objective to minimize so that I could see how the optimal spatial pattern of well locations changes as the number of wells used changes (you can imagine that the layout of 3 well system might be very different than the layout 6 well system).  You could also convert to using cost (installation plus O&M) and have a cost associated with "installing" any well location with a non-zero rate - this will incentivize MOU to use less wells...
